python/dgl/backend/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,356 @@
+
+            if self._rank is None:
+                self._rank = emb.rank
+                self._world_size = emb.world_size
+            else:
+                assert self._rank == emb.rank, 'MultiGPU rank for each embedding should be same.'
+                assert self._world_size == emb.world_size, 'MultiGPU world_size for each embedding should be same.'
+            if self._rank <= 0:
+                # may overflow if steps > 2B
+                emb_name = emb.name
+                state_step = create_shared_mem_array(emb_name+'_step', (emb.emb_tensor.shape[0],), th.int).zero_()
+                state_mem = create_shared_mem_array(emb_name+'_mem', emb.emb_tensor.shape, th.float32).zero_()
+                state_power = create_shared_mem_array(emb_name+'_power', emb.emb_tensor.shape, th.float32).zero_()
+            if self._rank == 0:
+                state = (state_step, state_mem, state_power)
+                emb_name = emb.name
+                for i in range(1, self._world_size):
+                    # send embs
+                    emb.store.set(emb_name+'_opt', emb_name)
+            elif self._rank > 0:
+                # receive
+                emb_name = emb.name
+                emb.store.wait([emb_name+'_opt'])
+                state_step = get_shared_mem_array(emb_name+'_step', (emb.emb_tensor.shape[0],), th.int)
+                state_mem = get_shared_mem_array(emb_name+'_mem', emb.emb_tensor.shape, th.float32)
+                state_power = get_shared_mem_array(emb_name+'_power', emb.emb_tensor.shape, th.float32)
+
+            state = (state_step, state_mem, state_power)
+            emb.set_optm_state(state)
+
+    def update(self, idx, grad, emb):
+        """ Update embeddings in a sparse manner
+        Sparse embeddings are updated in mini batches. we maintains gradient states for
+        each embedding so they can be updated separately.
+
+        Parameters
+        ----------
+        idx : tensor
+            Index of the embeddings to be updated.
+        grad : tensor
+            Gradient of each embedding.
+        emb : NodeEmbedding
+            Sparse embedding to update.
+        """
+        with th.no_grad():
+            beta1 = self._beta1
+            beta2 = self._beta2
+            eps = self._eps
+
+            clr = self._lr
+            state_step, state_mem, state_power = emb.optm_state
+            exec_dev = grad.device
+            state_dev = state_step.device
+
+            # the update is non-linear so indices must be unique
+            grad_indices, inverse, cnt = th.unique(idx, return_inverse=True, return_counts=True)
+            state_idx = grad_indices.to(state_dev)
+            state_step[state_idx] += 1
+            state_step = state_step[state_idx].to(exec_dev, non_blocking=True)
+            orig_mem = state_mem[state_idx].to(exec_dev, non_blocking=True)
+            orig_power = state_power[state_idx].to(exec_dev, non_blocking=True)
```

Comment:
do you mean the data to GPU to perform the operations in Adam and move them back to CPU? how much can it speed up?
`non_blocking` isn't very useful if the tensors are not stored in pinned memory originally.

Reply:
Tensors are stored in pinned memory here.